fmc/src/flow/rt_alias.rs

```diff
@@ -11,40 +11,168 @@ Abstract:
 
 #[derive(Default)]
 pub struct RtAliasLayer {}
 
 impl DiceLayer for RtAliasLayer {
     /// Perform derivations for the DICE layer
-    fn derive(_env: &FmcEnv, _input: &DiceInput) -> CaliptraResult<DiceOutput> {
-        // TODO : implement derivation.
-        Err(CaliptraError::FMC_RT_ALIAS_UNIMPLEMENTED)
+    fn derive(
+        env: &mut FmcEnv,
+        hand_off: &HandOff,
+        input: &DiceInput,
+    ) -> CaliptraResult<DiceOutput> {
+        // Derive CDI
+        let cdi = *okref(&Self::derive_cdi(env, hand_off, input.cdi))?;
+
+        // Derive DICE Key Pair from CDI
+        let key_pair = Self::derive_key_pair(env, cdi, input.subj_priv_key)?;
+
+        // Generate the Subject Serial Number and Subject Key Identifier.
+        //
+        // This information will be used by next DICE Layer while generating
+        // certificates
+        let subj_sn = X509::subj_sn(env, &key_pair.pub_key)?;
+        let subj_key_id = X509::subj_key_id(env, &key_pair.pub_key)?;
+
+        let output = input.to_output(key_pair, subj_sn, subj_key_id);
+
+        // Generate Local Device ID Certificate
+        Self::generate_cert_sig(env, input, &output)?;
+        Ok(output)
     }
 }
 
 impl RtAliasLayer {
     #[inline(never)]
     pub fn run(env: &mut FmcEnv, hand_off: &HandOff) -> CaliptraResult<()> {
-        cprintln!("[art] Extend PCRs");
+        cprintln!("Extend PCRs");
         Self::extend_pcrs(env, hand_off)?;
-        Ok(())
+
+        // Retrieve Dice Input Layer from Hand Off and Derive Key
+        match Self::dice_input_from_hand_off(hand_off) {
+            Ok(input) => {
+                let _ = Self::derive(env, hand_off, &input);
+                Ok(())
+            }
+            _ => Err(CaliptraError::FMC_RT_ALIAS_DERIVE_FAILURE),
+        }
+    }
+
+    /// Retrieve DICE Input from HandsOff
+    ///
+    /// # Arguments
+    ///
+    /// * `hand_off` - HandOff
+    ///
+    /// # Returns
+    ///
+    /// * `DiceInput` - DICE Layer Input
+    fn dice_input_from_hand_off(hand_off: &HandOff) -> CaliptraResult<DiceInput> {
+        // Create initial output
+        let input = DiceInput {
+            cdi: hand_off.fmc_cdi(),
+            subj_priv_key: hand_off.fmc_priv_key(),
+            auth_key_pair: Ecc384KeyPair {
+                priv_key: KeyId::KeyId5,
+                pub_key: Ecc384PubKey::default(),
+            },
+            auth_sn: [0u8; 64],
+            auth_key_id: [0u8; 20],
+            uds_key: hand_off.fmc_cdi(),
+        };
+
+        Ok(input)
     }
 
     /// Extend current and journey PCRs
     ///
     /// # Arguments
     ///
     /// * `env` - FMC Environment
+    /// * `hand_off` - HandOff
     pub fn extend_pcrs(env: &mut FmcEnv, hand_off: &HandOff) -> CaliptraResult<()> {
         extend_current_pcr(env, hand_off)?;
         extend_journey_pcr(env, hand_off)?;
         Ok(())
     }
+
+    /// Permute Composite Device Identity (CDI) using Rt TCI and Image Manifest Digest
```

Comment:
Could you add something to the comment to describe that RT Alias CDI overwrites the FMC Alias CDI? (Unless I'm misunderstanding how this works)

Reply:
Updated